### PR TITLE
Remove docker entirely

### LINKED ISSUE #2595: Remove support for Kubernetes 1.23 and Docker
### Description of the feature you would like to add / User story

Kubernetes 1.23 is already in maintenance mode and is supposed to reach EOL in February. That said, we can perhaps drop support for 1.23 from KubeOne 1.6. Once we do that, we can also drop support for Docker because support for dockershim has been removed in 1.24.

### Solution details

- Increase the minimum Kubernetes version to 1.23
- Start removing support for Docker from the code base

### PR DESCRIPTION
**Pull request by kron4eg**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2595

**What type of PR is this?**
/kind cleanup

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
* Support for docker is removed
  * containerRuntime.docker became a no-op.
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

File: pkg/apis/kubeone/helpers_test.go [hunk 1]
```diff
@@ -75,83 +75,6 @@ func TestFeatureGatesString(t *testing.T) {
 	}
 }
 
-func TestContainerRuntimeConfig_MachineControllerFlags(t *testing.T) {
```
Comment (xmudrii) [inline pkg/apis/kubeone/helpers_test.go:78]: This is also testing containerd, why are we removing this test?

File: pkg/apis/kubeone/validation/validation.go [hunk 2]
```diff
@@ -345,35 +336,6 @@ func ValidateKubernetesSupport(c kubeoneapi.KubeOneCluster, fldPath *field.Path)
...
-
-	var found bool
-	for _, x := range allCRs {
-		if !reflect.ValueOf(x).IsNil() {
-			if found {
-				allErrs = append(allErrs, field.Invalid(fldPath, x, "only 1 container runtime can be activated"))
-			}
-			found = true
-		}
-	}
```
Comment (xmudrii) [inline pkg/apis/kubeone/validation/validation.go:364]: I'm wondering if leaving such a function might be useful if we decide to add support for another CRI (e.g. CRI-O)

File: pkg/state/cluster.go [hunk 3]
```diff
@@ -282,7 +281,7 @@ func (h *Host) Initialized() bool {
 
 // IsProvisioned checks are CRI and Kubelet provisioned on a host
 func (h *Host) IsProvisioned() bool {
-	return (h.ContainerRuntimeDocker.IsProvisioned() || h.ContainerRuntimeContainerd.IsProvisioned()) && h.Kubelet.IsProvisioned()
+	return (h.ContainerRuntimeContainerd.IsProvisioned()) && h.Kubelet.IsProvisioned()
```
Comment (xmudrii) [inline pkg/state/cluster.go:284]: Very tiny nit:
```suggestion
	return h.ContainerRuntimeContainerd.IsProvisioned() && h.Kubelet.IsProvisioned()
```

File: pkg/apis/kubeone/helpers_test.go (hunk 1 above)
Comment (kron4eg) [reply pkg/apis/kubeone/helpers_test.go:78]: "Collateral damage". Probably I got too excited to remove docker and took some extra :shrug: 

Reverted back to original.

File: pkg/state/cluster.go (hunk 3 above)
Comment (kron4eg) [reply pkg/state/cluster.go]: Fixed :+1:

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve

**Pull request merged**